python/cucumber_expressions/parameter_type.py

```diff
@@ -0,0 +1,79 @@
+from __future__ import annotations
+
+import re
+
+from cucumber_expressions.errors import CucumberExpressionError
+
+ILLEGAL_PARAMETER_NAME_PATTERN = r"([\[\]()$.|?*+])"
+UNESCAPE_PATTERN = r"(\\([\[$.|?*+\]]))"
+
+
+class ParameterType:
+    def check_parameter_type_name(self, type_name):
+        if not self.is_valid_parameter_type_name(type_name):
+            raise CucumberExpressionError(
+                f"Illegal character in parameter name {type_name}. Parameter names may not contain '[]()$.|?*+'"
+            )
+
+    @staticmethod
+    def is_valid_parameter_type_name(type_name):
+        return not bool(re.compile(ILLEGAL_PARAMETER_NAME_PATTERN).match(type_name))
+
+    def transform(self, group_values: list[str]):
+        return self.transformer(*group_values)
+
+    @staticmethod
+    def compare(pt1: ParameterType, pt2: ParameterType):
+        if pt1.prefer_for_regexp_match and not pt2.prefer_for_regexp_match:
+            return -1
+        if pt2.prefer_for_regexp_match and not pt1.prefer_for_regexp_match:
+            return 1
+        _a_name = len(pt1.name or "")
+        _b_name = len(pt2.name or "")
+
+        if _a_name < _b_name:
+            return -1
+        elif _a_name > _b_name:
+            return 1
+        else:
+            return 0
+
+    # Create a new Parameter
+    #
+    # @param name the name of the parameter type
+    # @param regexp [Array] list of regexps for capture groups. A single regexp can also be used
+    # @param type the return type of the transformed
```

Comment:
```suggestion
    # @param param_type the return type of the transformed
```

Reply:
I have changed param_type -> type, with the caveat that I'm not 100% sure of the consequences, but the tests were fine...

Reply:
@jsa34 @aurelien-reeves 
Using `type` as parameter name normally causes problems/warnings with static code analysis tools, like `pylint`.
REASON: `type()` is a builtin function in Python to determine the type of an object.

Therefore, using the `param_type` as parameter name is probably the better choice.
NOTE: I still need to look over the complete implementation as whole.